File: LayoutTests/http/tests/navigation/resources/back-send-referrer-helper.php
<?php
    // Prevent from being cached.
    header("Cache-Control: no-cache, private, max-age=0");
    header("Content-Type: text/html");
?>

<html>

<script>
    window.name = parseInt(window.name) + 1;
</script>

Referrer: <?php echo $_SERVER['HTTP_REFERER']; ?>
<br/>
window.name: <script>document.write(window.name)</script>

<form name=loopback action="" method=GET></form>

<script>
    if (window.name == 1) {
        // Navigate once more to add a history entry.
        document.loopback.submit();
    } else if (window.name == 2) {
        history.go(-1);
    } else {
        if (window.layoutTestController)
            window.layoutTestController.notifyDone();
    }
</script>

</html>

File: LayoutTests/http/tests/navigation/resources/back-send-referrer-helper.py
#!/usr/bin/env python3

import os
import sys

referer = os.environ.get('HTTP_REFERER', '')

sys.stdout.write(
    'Cache-Control: no-cache, private, max-age=0\r\n'
    'Content-Type: text/html\r\n\r\n'
    '\n'
    '<html>\n'
    '\n'
    '<script>\n'
    '    window.name = parseInt(window.name) + 1;\n'
    '</script>\n'
    '\n'
    'Referrer: {}\n'
    '<br/>\n'
    'window.name: <script>document.write(window.name)</script>\n'
    '\n'
    '<form name=loopback action="" method=GET></form>\n'
    '\n'
    '<script>\n'
    '    if (window.name == 1) {{\n'
    '        // Navigate once more (in a timeout) to add a history entry.\n'
    '        setTimeout(function() {{document.loopback.submit();}}, 0);\n'
    '    }} else if (window.name == 2) {{\n'
    '        history.go(-1);\n'
    '    }} else {{\n'
    '        if (window.testRunner)\n'
    '            window.testRunner.notifyDone();\n'
    '    }}\n'
    '</script>\n'
    '\n'
    '</html>\n'.format(referer)
)